src/shogun/mathematics/linalg/LinalgBackendBase.h

```diff
@@ -28,69 +27,63 @@
  * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
  * POSSIBILITY OF SUCH DAMAGE.
  *
- * Authors: 2016 Pan Deng, Soumyajit De, Viktor Gal
+ * Authors: 2016 Pan Deng, Soumyajit De, Heiko Strathmann, Viktor Gal
  */
 
 #include <shogun/lib/config.h>
-#include <shogun/io/SGIO.h>
+#include <shogun/lib/common.h>
 #include <shogun/lib/SGVector.h>
-#include <shogun/mathematics/linalg/BaseVector.h>
+#include <shogun/io/SGIO.h>
+#include <shogun/mathematics/linalg/GPUMemoryBase.h>
 #include <memory>
 
-#ifndef GPU_Vector_H__
-#define GPU_Vector_H__
-
-#ifdef HAVE_CXX11
+#ifndef Linalg_Backend_Base_H__
+#define Linalg_Backend_Base_H__
 
 namespace shogun
 {
 
-/** GPU vector wrapper structure */
-template <class T>
-struct GPUVector : public BaseVector<T>
+/** Base class for generic linalg methods */
+class LinalgBackendBase
 {
-	/** Structure for different types of GPU vectors */
-	struct GPUArray;
-
-	/** Opaque pointer of GPUArray */
-	std::unique_ptr<GPUArray> gpuarray;
-
-	/** Default Constructor */
-	GPUVector();
-
-	/** Wrap SGVector */
-	GPUVector(const SGVector<T> &vector);
-
-	/** Copy Constructor*/
-	GPUVector(const GPUVector<T> &vector);
+public:
+	/** Generic vector dot-product macro */
+	#define BACKEND_GENERIC_DOT(Type) \
+	virtual Type dot(const SGVector<Type>&, const SGVector<Type>&) const \
+	{  \
+		SG_SNOTIMPLEMENTED; \
```

Comment:
I think we should have a better error here.
Create a method in the base class that says "Linalg method is not (yet) implemented in the registered backend \"Vienna\". For that, you would need get_name. So can think about making the backends a subclass of CSGObject (in contrast to what I said above)